Remove effectively private classes from ABI

Tagakov opened the pull request:
These two code snippets should be identical from the ABI perspective, but they are not:

```kotlin
class Class {
  private class PrivateClass
}
```

```kotlin
class Class
```

This can't be fixed without breaking backwards compatibility with Java, since private classes could be referenced from Java. Additionally invisible companions are also not being removed from ABI, two code snippets below also are not identical from ABI perspective:

```kotlin
class Class {
  private companion object
}
```

```kotlin
class Class
```

Since the fix breaks backwards compatibility the change is hidden behind a configuration keys.

https://youtrack.jetbrains.com/issue/KT-64590/Effectively-private-classes-are-not-being-removed-from-ABI

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -35,8 +25,9 @@ enum class AbiMethodInfo {
 sealed class AbiClassInfo {
-    object Public : AbiClassInfo()
-    class Stripped(val methodInfo: Map<Method, AbiMethodInfo>) : AbiClassInfo()
+    object Keep : AbiClassInfo()
+    class Strip(val methodInfo: Map<Method, AbiMethodInfo>) : AbiClassInfo()
```
Could you please revert this rename for the time being. We don't usually perform meaningful changes and renames in the same commit, unless strictly necessary.

udalov commented on plugins/jvm-abi-gen/testData/compare/privateCompanionClasses/base/test.kt:
```diff
@@ -0,0 +1,5 @@
+package test
+
+class Class {
+    private companion object
```
Instead of this separate test, please declare a companion object in `nestedPrivateClasses`.

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt:
```diff
@@ -53,6 +53,11 @@ class JvmAbiOutputExtension(
         val removeDebugInfo: Boolean,
     ) : OutputFileCollection {
+
+        private val classesToBeDeleted = abiClassInfos.mapNotNullTo(mutableSetOf()) { (className, action) ->
```
Please inline this variable

Tagakov replied:
I assume that you ask for that because of better readability and cognitive scope management. 

I can do it, but originally I put that property that high because I wanted to compute a set of removed classes only once per `jvm-abi-gen` invocation If I inline the property it will be computed for every class in the resulting jar which could be possibly a lot.

As a compromise I made it immutable so nobody think that it is being changed elsewhere. If I am wrong and there is another reason for inlining - please let me know!

udalov replied:
You're right, thanks for the explanation!

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -149,9 +152,11 @@ private fun AnnotationVisitor.visitKotlinMetadata(header: Metadata) {
     constructors.removeIf { it.visibility.isPrivate }
     (this as KmDeclarationContainer).removePrivateDeclarations()
+    nestedClasses.removeIf { "$name\$$it" in classesToBeDeleted }
+    companionObject = companionObject?.takeIf { "$name\$$it" !in classesToBeDeleted }
```
```suggestion
    companionObject = companionObject.takeUnless { "$name\$$it" in classesToBeDeleted }
```

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt:
```diff
@@ -69,13 +74,15 @@ class JvmAbiOutputExtension(
                     // Note that outputFile may be null, e.g., for empty $DefaultImpls classes in the JVM backend.
                     outputFile == null ->
                         null
```
Let's move it one line above into `if (outputFile == null) return@mapNotNull null`, and make the `when` exhaustive via `when (abiInfo) ...`

udalov reviewed (changes requested):
Private nested classes (including private companion objects) cannot be access from Java. The Java interop matter is relevant only for private top-level classes. So I'd suggest to always remove private nested classes, and remove top-level classes if the option is enabled. And then the option should probably be named `removePrivateTopLevelClasses`.

Tagakov commented:
I slightly refactored my implementation, I think now it is easier to read.

artem-zinnatullin commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/overrides/IrVisibilityUtil.kt:
```diff
@@ -16,13 +15,12 @@ fun isVisibleForOverride(overriding: IrOverridableMember, fromSuper: IrOverridab
 
     return when {
-        visibility.isNonPrivate() -> parentClassOrNull?.isEffectivelyPrivate() ?: false
+        isNonPrivate -> parentClassOrNull?.isEffectivelyPrivate() ?: false
```
My god, these double negations with fallbacks to `null == false` are quite bad. Just why…

Tagakov replied:
It ain't owe nothing to nobody!

artem-zinnatullin commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -64,31 +56,39 @@ sealed class AbiClassInfo {
         override fun defineClass(
             version: Int, access: Int, name: String, signature: String?, superName: String, interfaces: Array<out String>
         ) {
             // Always keep annotation classes
```
What if I change kdoc on the annotation, it would shift source lines, would that break ABI? 

Sounds like annotation should also go through debug info stripping instead of just keeping it as is all the time, I might be wrong, haven't checked.

Tagakov replied:
Checked it, debug info doesn't affect annotation. I think it happens due to the fact that they can't have members at all.

artem-zinnatullin commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -154,14 +155,16 @@ class JvmAbiClassBuilderInterceptor : ClassGeneratorExtension {
-                !localOrAnonymousClass && !isWhenMappingClass -> {
+            abiClassInfo[internalName] = when {
+                keepEverything -> AbiClassInfo.Public
+                classRemovedFromAbi -> AbiClassInfo.Delete
```
```suggestion
                removeClassFromAbi -> AbiClassInfo.Delete
```

or similar naming would be more correct it seems

artem-zinnatullin commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -171,3 +174,9 @@ class JvmAbiClassBuilderInterceptor : ClassGeneratorExtension {
+
+private fun shouldBeRemovedFromAbi(irClass: IrClass?, removePrivateTopLevelClasses: Boolean) = when {
+    irClass == null -> false
+    DescriptorVisibilities.isPrivate(irClass.visibility) -> removePrivateTopLevelClasses || !irClass.isTopLevel
```
Should be

```suggestion
    DescriptorVisibilities.isPrivate(irClass.visibility) -> (irClass.isTopLevel && removePrivateTopLevelClasses) || !irClass.isTopLevel
```

Tagakov replied:
I need a second opinion on that. For me the shorter version is easier to read.

udalov replied:
I'm also for the shorter version

artem-zinnatullin commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -32,7 +35,7 @@ fun abiMetadataProcessor(annotationVisitor: AnnotationVisitor): AnnotationVisito
             KotlinClassMetadata.transform(header) { metadata ->
                 when (metadata) {
                     is KotlinClassMetadata.Class -> {
-                        metadata.kmClass.removePrivateDeclarations()
```
So before this PR ABI always kept all inner classes, wut

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt:
```diff
@@ -202,7 +203,9 @@ class JvmAbiOutputExtension(
                 // Classes form a tree by nesting, so none of the children have been visited yet.
-                innerClassesByOuterName[next]?.mapNotNullTo(stack) { it.name.takeIf(abiClassInfos::contains) }
+                innerClassesByOuterName[next]?.mapNotNullTo(stack) { info ->
+                    info.name.takeIf { abiClassInfos[it] != AbiClassInfo.Deleted }
```
I would prefer to use `getValue` here as well, instead of `get`

udalov replied:
Fixed in https://github.com/JetBrains/kotlin/commit/2dfd6ad4077784846eb52cbb0e81abebe29975c8

Tagakov replied:
Thank you for the correction. Fail fast.

Tagakov reviewed (commented):
Looks good to me! Thank you for the fixups!

madsager commented:
Testing this change, I found a couple of issues that are going to break us if this lands as is.

---

Issue 1:

Private interface can be used as super-types for public classes. When that happens, both kotlinc and javac require the private interface to be present in the abi jar:

```
class Outer {
  // Cannot strip this local private interface from abi jar.
  private interface I
  public class A: I
}
```

We could just not strip private interfaces as a quick fix for that. The alternative is to trace through the super chain and mark all interfaces encountered as required in the output.

---

Issue 2:

It looks like the current implementation expects all inner classes that it encounters to be from the current compilation unit and therefore have an entry in `abiClassInfos`. When an inner class from a library dependency is encountered that throws an exception. We can probably fix that by using `abiClassInfos[it]` instead of the throwing `abiClassInfos.getValue(it)` when adding inner classes?

---

Issue 3:

We perform annotation processing on abi jars. In places we have patterns such as:

```
class Outer {
  @MyAnno
  fun foo() { }
  private object Annotations {
    annotation class MyAnno
  }
}
```

Since the private object class is now stripped, that means that we cannot find the annotation which is present on a public member. This seems somewhat similar to (but more obscure than) the private interface case. We are leaking out something private on a public member. It is not clear to me if we should trace through this and keep all annotations on public elements and their surrounding context. That would probably be prudent to make the ABI jar consistent in the sense that things reachable from public declarations need to stay.

---

I'm still looking through a few more issues, but these were the biggest offenders.

Tagakov commented:
Hi @madsager, thanks for your attention!

Issue 1: It is relatively simple and, at the first glance, it doesn't seem to be hard to leave `private interface`s in place if they are being implemented by some public entity. It should not affect the ABI stability, because no matter what implementing the interface forces a user to have all members coming from the interface to be public. Meaning that even if we somehow manage to safely strip out the private interface any change to it is guaranteed to change the ABI in your case. I can give it a shot if you think it is worth fixing, @udalov please verify that it is ok for you as well.

Issue 2: Can you please provide an example of your case? I can't understand how you can have an inner class in code you own coming from an external library. The change you're proposing seems reasonable, but I am curious what causes the issue and would prefer to write a test for it if it is possible.

Issue 3: The adhoc fix is similar to the Issue 1 but way more complex since an annotation could be put in a lot of places from plainly annotated class to obscure generic type somewhere in typealias or function returning type. I don't think I'll be able to catch all possible cases on the first try and anyways it intuitively seems like a proper analysis requires significant work computation-wise so I don't believe it is the right way to go. Annotation classes are already being kept unconditionally I wonder if the problem is with the fact that annotation is nested or outer class is being removed. I'll check and return with the results.

Preliminarily it seems like there is no easy fix for Issue 3 and chances are that this PR breaking backwards compatibility for others seems to be higher than I initially anticipated. Additionally it looks like @madsager has more issues with the change and I suspect he could be not the only one, unfortunately :( @udalov, what do you think about introducing a configuration flag `strictAbi` or a similar one that will only enable abi-gen improvements (including internal removal) for the ones who set it?

Looking forward to your answers, folks!

___

Issue 1, Issue 3 offtopic: I think you're just relying on a bug in jvm-abi-gen and would prefer you (and everyone else, yeah...) to migrate from it. I understand that it is not how backwards compatibility works tho, but regardless of the results of this PR if you personally want to give the migration a shot, I'd be glad to help with it by providing a script that will try to do it for you automatically in IDE (assuming your project could be opened using IDEA). Just to make world a better place.

Another Issue 3 offtopic: I strongly believe that you should do your annotation processing against full jars instead of abi jars. Doing the opposite prevents you from getting benefits from more stable ABI, after this PR I am going to open another one which will strip (behind the configuration flag, ofc) `internal` members from abi.jar along with private ones, if your setup runs AP against abi.jar or probably compiles your tests against abi.jar instead of full jar you won't be able adopt it unfortunately :( But even in the scope of current PR you'll have problems with your setup if you try to do something like this:
```kotlin
class Public {
  fun publicBridge() = Generated_SomeEntryPointForAP()
  @SomeAnnotation
  private class SomeEntryPointForAP {...}
}
```

madsager commented:
Sorry for the slow reply here. I was having a bit of trouble reproducing issue 2. It is based on auto-generated code and it took me a while to figure out that it is actually caused by a name conflict in file classes. That is a little annoying, but since this compiles fine, we should probably make sure that we guard against it. Repro:

```
$ cat lib/A.kt
package lib

class A
[ager@contractum] /tmp/repro 
$ cat lib/AKt.kt 
package lib

object AKt {
    class B {
        companion object {
            fun create() = B()
        }
    }
}

inline fun A.foo() = AKt.B.create()

$ cat use/A.kt
package lib

fun A.extension() = foo()

$ kotlinc lib/A.kt lib/AKt.kt -d lib/lib.jar -Xplugin=jvm-abi-gen.jar -P plugin:org.jetbrains.kotlin.jvm.abi:outputDir=lib/lib-header.jar
lib/AKt.kt:11:1: warning: expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types
inline fun A.foo() = AKt.B.create()
^

$ kotlinc use/A.kt -d use/use.jar -cp lib/lib-header.jar -Xplugin=jvm-abi-gen.jar -P plugin:org.jetbrains.kotlin.jvm.abi:outputDir=use/use-header.jar
exception: java.util.NoSuchElementException: Key lib/AKt$B is missing in the map.
	at kotlin.collections.MapsKt__MapWithDefaultKt.getOrImplicitDefaultNullable(MapWithDefault.kt:24)
	at kotlin.collections.MapsKt__MapsKt.getValue(Maps.kt:360)
	at org.jetbrains.kotlin.jvm.abi.JvmAbiOutputExtension$AbiOutputFiles.addInnerClasses(JvmAbiOutputExtension.kt:191)
	at org.jetbrains.kotlin.jvm.abi.JvmAbiOutputExtension$AbiOutputFiles.access$addInnerClasses(JvmAbiOutputExtension.kt:53)
	at org.jetbrains.kotlin.jvm.abi.JvmAbiOutputExtension$AbiOutputFiles$asList$classFiles$1$1.visitEnd(JvmAbiOutputExtension.kt:162)
        ...
```

My two cents: I think we should definitely fix issue 1 as otherwise you cannot use jars produced by jvm-abi-gen to compile against for valid Kotlin code neither with javac or kotlinc. Issue 2 is annoying, but I think we should fix it as well. Issue 3 I'm still not sure what I think about. :-\

madsager commented:
PS: Keeping all interfaces, guarding against the missing map entry, and manually changing visibility for companion objects with annotation defined inside makes everything pass on my side with the default setting of only stripping private nested classes. Based on what I have seen so far, I think we should make liberal use of flags for settings here. This apparently breaks us in subtler ways that I had thought. Even the sorting change is breaking us because annotation processors and processing order of declarations, so I would prefer that we put in a flag even for that one as well (default sorting on, but allow to turn it off).

udalov commented:
@Tagakov Let's fix the issue 1 and 2, put all the new behavior under the same flag, and rename the flag to `removePrivateClasses`.

Tagakov commented:
@madsager added option to disable members sorting: https://github.com/JetBrains/kotlin/pull/5252

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -58,8 +61,15 @@ sealed class AbiClassInfo {
+    private val removePrivateClasses: Boolean,
 ) : ClassGeneratorExtension {
-    val abiClassInfo: MutableMap<String, AbiClassInfo> = mutableMapOf()
+    private var abiClassInfoBuilder = AbiClassInfoBuilder()
+
+    fun buildAbiClassInfoAndReleaseResources(): Map<String, AbiClassInfo> {
+        return abiClassInfoBuilder.buildClassInfo().also {
+            abiClassInfoBuilder = AbiClassInfoBuilder()
+        }
+    }
```
I'm storing more info about classes to be able to keep private parents of annotations and exposed private interfaces (their parents too, if any). Additionally I get better encapsulation == easier to read/follow the code. Please make sure it is ok approach.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -146,36 +162,118 @@ class JvmAbiClassBuilderInterceptor(
         }
 
+        override fun visitInnerClass(name: String, outerName: String?, innerName: String?, access: Int) {
+            abiClassInfoBuilder.innerClassToParent(name, outerName)
```
Wasn't able to find a better way to record class hierarchy. I think I can achieve the same result using `irClass` let me know which approach is better.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -146,36 +162,118 @@ class JvmAbiClassBuilderInterceptor(
+
+private fun shouldRemoveFromAbi(irClass: IrClass?, removePrivateClasses: Boolean): Boolean = when {
+    irClass == null -> false
+    irClass.isFileClass -> false
```
This was necessary to keep machinery used for multifile classes working. They have package-private visibility as default which is treated as private by `isEffectivelyPrivate` function, I like the idea of stripping out as many classes from ABI as possible so I went this route instead of writing my on `isEffectivelyPrivate` function that will treat `package` visibility as non-private.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -146,36 +162,118 @@ class JvmAbiClassBuilderInterceptor(
+private fun shouldRemoveFromAbi(irClass: IrClass?, removePrivateClasses: Boolean): Boolean = when {
+    irClass == null -> false
+    irClass.isFileClass -> false
+    removePrivateClasses -> !irClass.isTopLevel && irClass.isEffectivelyPrivate()
```
Note the `!irClass.isTopLevel` here, it is added only for the sake of testing the changes by @madsager since it won't be possible to build his project otherwise.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -18,6 +18,8 @@ import org.jetbrains.org.objectweb.asm.Opcodes
     annotationVisitor: AnnotationVisitor,
     removeDataClassCopyIfConstructorIsPrivate: Boolean,
+    classesToBeDeleted: Set<String>,
+    pruneClass: Boolean,
```
The idea behind this flag is to get rid of anything in the class that was about to be deleted but left in place because of nested private interface or annotation.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -32,13 +34,15 @@ fun abiMetadataProcessor(
                     is KotlinClassMetadata.Class -> {
-                        metadata.kmClass.removePrivateDeclarations(removeDataClassCopyIfConstructorIsPrivate)
+                        //the single reason for pruning is to leave public class in abi.jar without any content in it
+                        if (pruneClass) metadata.kmClass.visibility = Visibility.PUBLIC
```
This line is covered with tests. If removed kotlinc refuses to work with annotation from a private class in an external library. I am not sure it is the best idea, maybe we can just delete this if we only expect those annotations to be read by KSP/javac.

madsager replied:
A related issue that I found a couple of occurrences of:

```
private class MyClass

@MyAnnotation(MyClass::class.java)
fun publicFun() = ...
```

Again, this is used for annotation processing. My guess is that the goal is to communicate information to an annotation processor without making the classes carrying that information part of the ABI in the end. Would it be simple to trace through annotation values as well and keep any classes referenced from annotations?

Since everything here is under a flag, I would also be OK with landing without keeping annotations and private classes reachable from anntotations. Then we can add a flag for doing that at a later point. For now, we can just not use it (or make things public if that is acceptable for the teams doing this).

Tagakov replied:
> we would allow something to compile with ABI jars that would not compile with the full jar
Agree 100%, my bad! Got dragged to a bad decision wanting to write a test somewhat resembling your case. :) Will remove that and look what happens.

> My guess is that the goal is to communicate information to an annotation processor without making the classes carrying that information part of the ABI in the end.

This is a tough one, it is relatively easy to check "typical" places for annotations like methods and classes, not sure about perf impact from that tho (/cc @udalov) but what bothers me is a depth of the rabbit hole =\ @madsager, let's say we figure out that we need to keep `private class MyClass` in the abi.jar what extra information from that class also needs to be kept? It's methods, interfaces, super(recursively?!), inner classes, etc? 

Kotlin allows having private/internal members in annotation arguments no matter what and, I guess, we have to live with it. But what you think is appropriate to keep in such leaked classes? I vote for "nothing" but not sure if it's acceptable. I can see a usecase for that right away: having JSR-330's `@Qualifier` with private class as a key like: `@NoMatterPrivateOrPublicQualifier(PrivateClassAsAKey::class)` so at least we need to keep the private class referenced here. As for other information - I wish we could just make it the problem of the annotation processor author :)

___

@udalov can you please share your thoughts on this one? I think I'm starting to get analysis paralysis here >_<

madsager replied:
Yes, I agree. I'm leaning towards not trying to fix the annotation issues caused by us doing annotation processing on ABI jars and incorrectly assuming that an annotation processor can inspect non-ABI members in an ABI jar. I think jvm-abi-gen should maintain what is public ABI for compilation with javac and kotlinc. I'm not aware of any cases where javac and kotlinc rely on private annotations or private class constants in annotations for compilation. Which could argue in favor of stripping private annotations from ABI jars and having annotation processor authors who want to use ABI jars deal with the fallout of that (the fallout is small on my end).

udalov replied:
I agree with Mads about not fixing the annotation issue. If the flag is enabled, private annotation classes should be removed just as other private classes.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -149,24 +153,27 @@ private fun AnnotationVisitor.visitKotlinMetadata(header: Metadata) {
+        copyFunShouldBeDeleted(removeCopyAlongWithConstructor),
+        pruneClass,
     )
+    nestedClasses.removeIf { "$name\$$it" in classesToBeDeleted }
```
Not sure if we need to be smarter here for pruned classes. Probably not, since all nested classes of a pruned should be considered effectively private.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt:
```diff
@@ -92,15 +92,27 @@ class JvmAbiOutputExtension(
+                                superName: String?,
+                                interfaces: Array<out String>?,
+                            ) {
+                                val prunedAccess = if (prune) access or Opcodes.ACC_PUBLIC else access
```
Same as in `@Metadata`. Not sure if I should fallback to package-private here instead.

Tagakov commented:
Hi everyone! 

Issue 2: @madsager Thank you for repro steps, I almost got it by myself, but my (wrong) assumption that it is not possible to compile a class while having the one with the same FQCN in the classpath led me astray, so it doesn't count :) It was super entertaining to learn!

I distilled your example and added it as a test to make it future proof. If you're curious about it, here is the link to the test: [click](https://github.com/JetBrains/kotlin/pull/5234/files#diff-8d26a9718559e3c8b19e7fdce938a8da6229a6981cc25adad54670ecbd039abd)

___

Issue 1 and Issue 3: Given that annotations already have special handling in `jvm-abi-gen` I decided to leave them in place unconditionally. Nonetheless if one wants to improve their ABI stability and they can afford it, they could always make their annotations' retention to be SOURCE and then those annotations won't be in the abi.jar. 

In the light of everything said above, the Issue 3 becomes a simplified version of the Issue 1. Meaning that we need to solve three things:

- Collect all annotations and publicly exposed private or, more important, effectively private interfaces;
- Keep everything from the previous step in abi.jar **plus** all their outer classes;
- Remove everything from the outer classes from the previous step except artificially kept inner classes.

```kotlin
//no class could be removed in this example
class PublicClass {
  class ExposingClass: KeptPrivateInterface

  private interface KeptPrivateInterface: KeptNestedPrivateClass.KeptEffectivelyPrivatePublicInterface

  private class KeptNestedPrivateClass {
    interface KeptEffectivelyPrivatePublicInterface
    
    fun publicFunThatShouldNotEndInTheAbi() = Unit
  }
}
```

This logic should apply whether we keep private classes in abi.jar or remove them.

___

@udalov I want to reiterate on putting all the new behavior under a single flag. I understand that we don't want to accidentally break someone's project. But I think that providing clients with "all-or-nothing" flag is not the best way. I believe that the number of people who can flip the flag and get all private classes removed is significantly smaller than the ones who can get away with only non-top-level private classes to be removed and @madsager is one of them. 

Maybe we can consider making the flag to be enum instead of boolean? 

In the meantime I moved all the changes under the single `removePrivateClasses` flag as you asked.

___

@madsager can I ask you to test the changes once more? I believe that all the issues you brought are fixed and really looking forward to hear that the new `jvm-abi-gen` becomes functional in your project(s). To test the changes you need to flip `removePrivateClasses` flag, note that the last commit in the PR disables removal of top-level classes (/cc @udalov) while making some tests fail. It was done only for testing purposes to be able to get more intel on other issues that this PR could possibly have in your project.

madsager reviewed (commented):
I'll give these changes a spin. A couple of comments while I wait for those results. :)

madsager commented:
> I'll give these changes a spin. A couple of comments while I wait for those results. :)

Results are in and it is looking good on my end.

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -86,11 +98,10 @@ class JvmAbiClassBuilderInterceptor(
+            keepClassAsIs = keepClassAsIs || access and Opcodes.ACC_ANNOTATION != 0
 
             internalName = name
+            superInterfaces = interfaces.asList()
```
Initially I tried to avoid a unnecessary copy here since we're going to copy the array to the builder at the end of the class processing anyways. Did you do it only as a defensive approach here?

udalov replied:
I did it only because using lists is preferable to using arrays in Kotlin. It's not a copy, `asList` creates a view, which is supposed to be cheap.

Tagakov replied:
Read it as `toList()` my bad. That makes total sense!

Tagakov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassInfoBuilder.kt:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    private fun keepInterface(className: String, visited: MutableSet<String>) {
+        if (!visited.add(className)) return
```
I am not sure if the memory footprint is a concern here, but just in case the `visited` set could be replaced by the `abiClassInfo[className] != AbiClassInfo.Public` check here. Although it will make the pattern less recognizeable by a future reader.

udalov replied:
Yes, I feel that `visited` is a bit more explicit.

Tagakov commented:
Hurray!

@madsager - its ready for a final test if you don't mind :)
@udalov - its ready for merge after tests :)

udalov commented:
I've discussed it with @madsager and decided to merge this as is for now.

Merged in 11d3ead9756dd0d8f4742a4c416b0f343891f9fa. Thank you for the contribution!
